### PR TITLE
calcite-backed SQL

### PR DESCRIPTION
**Pull request by rjullman**
Things that probably should still be done are:
(1) Add the sql shell to the atlas clis
(2) Made queries do range scans intelligently (current full range scans)

General note: Calcite is not well documented and often it is unclear what interfaces to implement to get desired functionality.

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/1393)

### CODE REVIEW

File: atlasdb-calcite/build.gradle [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    compile project(':atlasdb-dagger')
+    compile project(':atlasdb-service')
+    compile group: 'io.airlift', name: 'airline', version: '0.7'
+    compile(group: 'org.apache.calcite', name: 'calcite-core') {
+        exclude group: 'com.google.protobuf', module: 'protobuf-java'
+        exclude group: 'commons-lang', module: 'commons-lang'
+    }
+    compile group: "com.google.protobuf", name: "protobuf-java", version: libVersions.protobuf
+    compile group: "commons-lang", name: "commons-lang", version: libVersions.commons_lang
+    compile group: 'sqlline', name: 'sqlline', version: '1.1.9'
```
Comment (schlosna) [inline atlasdb-calcite/build.gradle]: curious why not [latest 1.2.0](https://github.com/julianhyde/sqlline/blob/master/HISTORY.md#120-2016-10-26) and should probably pull these versions out into the global version.props

**Comment by rjullman** [comment]
Let me lay out some more context here.

There have been a bunch of adhoc discussions about how people should inspect AtlasDB (outside of Java).  These discussions have mostly happened with @jboreiko @rhero @SerialVelocity @davidscottcohen (tagged here so they can correct me if I do not summarize faithfully).  Currently we have AtlasConsole (the replacement for the still internally used AtlasShell).  These both suffer from having a high barrier to entry for new users and so in practice only the atlas team and internal support know how to use them.  Consumers of atlas have historically asked for a SQL shell for atlas so that their general SQL knowledge transfers to inspect their data after they migrate to atlas.

Last hack week I implemented a jdbc driver from scratch (https://github.com/palantir/atlasdb/tree/feature/sql-updated) and it showed some promised, but ultimately seemed like a large undertaking to get working bugfree.  This PR is from this hack week and implements the interfaces to work with the Apache Calcite (https://calcite.apache.org/) JDBC driver.  In my opinion this is much more promising.  In the current state it works and is useful as a V0.  The only things I would add for a V1 are:
(1) integration into atlas cli, 
(2) tests that actually start a dropwizard server instead of the in memory hacks I added, and 
(3) implementation of range scans (instead of always doing full table scans).  
The first is straightforward, the second was hard for me during hack week but we already do things like this to test other parts of atlas (a docker guru will probably know what to do), and the third will mostly require trying to understand what interfaces to implement in the Calcite framework, which is largely under-documented.

tldr: This is SQL shell for atlas that will hopefully reduce barrier to entry and allow people to be comfortable migration to atlas.

File: atlasdb-calcite/src/main/java/com/palantir/atlasdb/calcite/AtlasColumnMetdata.java [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+            throw new RuntimeException("Row/column components cannot be deserialized in isolation."
+                    + "Use EncodingUtils.fromBytes() instead.");
+        }
+        switch (format()) {
+            case PROTO:
+                if (isComponent()) {
+                    throw new IllegalStateException("Row/column components cannot contain protobufs (invalid metadata).");
+                } else if (isNamedColumn()) {
+                    return ProtobufDeserializers.convertMessageToMap(
+                            column().get().getValue().hydrateProto(
```
Comment (rjullman) [inline atlasdb-calcite/src/main/java/com/palantir/atlasdb/calcite/AtlasColumnMetdata.java]: @gbonik pointed out that hydrateProto() requires the object class file, but convertMessageToMap does not use the class information.  It is definitely worth just loading the protobuf message directly without requiring the class file.

**Comment by rhero** [comment]
Hey @rjullman - are you okay with us closing this? We're probably not going to have time to fully review it, and given point (3) that you flag, we'd rather not have a feature that would trigger full table scans (and potentially causing heavy load on KVSs).

**Comment by gsheasby** [comment]
Closing, but leaving the branch around. Can resurrect this if @rjullman wants to work on it further.
